Steven/multicluster Split horizon multi network eds

Stevenjin8 opened the pull request:
**Please provide a description of this PR:**



**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

jaellio committed:
Initial changes to reach label selector from mesh config to update endpoints

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Pass MeshConfig to GlobalMergedWorkloadServicesCollection

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Add cluster and listener service filtering for waypoints and
e/w gateways

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Respond to PR feedback

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Add more tests

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Fix incorrect waypoint logic. In ambient mode, e/w gateways and waypoints
are both node.Type Waypoints

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Remove unused ServicesForGateway

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Updated failing lint and add a release note

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56654)

my-git9 committed:
support displaying secret info in record message (#56662)

Signed-off-by: xin.li <xin.li@daocloud.io>

Stevenjin8 committed:
[REMOVE ME] enable multinetwork ambient by default

keithmattix commented on pilot/pkg/serviceregistry/kube/controller/ambient/networks.go:
```diff
@@ -272,7 +272,7 @@ func remoteK8sGatewayToNetworkGateways(clusterID cluster.ID, gw *v1beta1.Gateway
 	gateways := []NetworkGateway{}
 	source := config.NamespacedName(gw)
-	for _, addr := range gw.Spec.Addresses {
+	for _, addr := range gw.Status.Addresses {
```
Ah, nice catch

keithmattix commented on pilot/pkg/config/kube/gateway/deploymentcontroller.go:
```diff
@@ -202,7 +202,7 @@ func getClassInfos() map[gateway.GatewayController]classInfo {
 			description:        "The default GatewayClass for Istio East West Gateways",
 			templates:          "waypoint",
-			disableNameSuffix:  true,
+			disableNameSuffix:  false,
```
Why do we want to keep the suffix? I'm not opposed to it or blocking, but curious about the thoguht process

Stevenjin8 replied:
We use the `GatewaySA` to set `NetworkGateway.ServiceAccount`

https://github.com/istio/istio/blob/c48b097b442c721ad526fce8d0e91aa124cf1037/pilot/pkg/serviceregistry/kube/conversion.go#L229-L231

which adds the suffix. However, the deployment controller reads directly from the `Gateway` object. if we do `disableNameSuffix: true` then the gateway pods will use `xxx` as the deployment name and service account, but WDS will advertise `xxx-east-west-gateway` as the service account name.

keithmattix replied:
Hm wonder how it works with waypoints then...in any case, this is fine for now

Stevenjin8 committed:
Also add Service SANS

zirain committed:
fix clustertrustbundles rbac (#56675)

zirain committed:
fix clusterTrustBundle template (#56676)

my-git9 committed:
support displaying traffic type for waypoint list (#56521)

* support displaying traffic type for waypoint list

Signed-off-by: xin.li <xin.li@daocloud.io>

* add unit test

Signed-off-by: xin.li <xin.li@daocloud.io>

---------

Signed-off-by: xin.li <xin.li@daocloud.io>

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56663)

mike-weiner committed:
Enhance genHTMLFragment to Include Anchor Tags in Table Output (#56645)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#56683)

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56684)

Stevenjin8 committed:
Move number theory functions to math module

Stevenjin8 committed:
Even capacity distribution with multiple gateways

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#56688)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#56693)

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56694)

hanxiaop committed:
Fix analysis having k8s system namespaces (#56297)

* fix analysis having k8s system namespaces

* add releasenote

istio-testing committed:
Update BASE_VERSION to master-2025-06-19T19-00-49 (#56699)

jaellio committed:
Filter services in ServiceServiceBuilder and check feature
flag for ambient multi network.

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#56708)

ramaraochavali committed:
minor changes to krt (#56647)

* wip

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix recomputeMu docs

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

---------

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56685)

jaellio committed:
Move feature flag set to avoid race

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

jaellio committed:
Fix race

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

istio-testing committed:
Automator: update go-control-plane in istio/istio@master (#56724)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56725)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56726)

AritraDey-Dev committed:
support ClusterTrustBundle to v1beta1 API (#56727)

* refactor: migrate ClusterTrustBundle to v1beta1 API

Fixes #56306

Signed-off-by: Aritra Dey <adey01027@gmail.com>

* added release notes

Signed-off-by: Aritra Dey <adey01027@gmail.com>

* modified metadata.yaml

Signed-off-by: Aritra Dey <adey01027@gmail.com>

---------

Signed-off-by: Aritra Dey <adey01027@gmail.com>

Stevenjin8 committed:
Automatically include waypoint svc of global svcs

* Some refactoring too

adam-cattermole committed:
wasmplugin: Allow configuration of fail reload strategy (#55604)

Signed-off-by: Adam Cattermole <acatterm@redhat.com>

darth-drew committed:
Fix prepetual out of sync status within argocd (#56695)

* Update daemonset.yaml

https://github.com/istio/istio/pull/55955 for reference. Fixes out of sync status within argocd

* change GOMEMLIMIT and GOMAXPROCS divisor from 0 to 1

* Add release notes

jewertow committed:
tls: add PQC compliance policy (#56331)

* tls: add PQC compliance policy

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Remove max TLS when compliance policy is PQC

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Set cipher suites explicitly in PQC mode

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Update documentation for 'pqc' compliance policy

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Add a release note

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

---------

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#56744)

Stevenjin8 committed:
Remove unused multiCluster values (#56736)

nicole-lihui committed:
revert istio-remote service when profil=remote (#56524)

* revert istio-remote service when profil=remote

closed #56328

* fix helm values

* fix release note

* fix not install istiod

* iop add EnabledLocalInjectorIstiod & fix configCluster & add ut

* change release note bug to feature

AritraDey-Dev committed:
feat: better istioctl ps output (#56441)

* feat: better istioctl ps output

* feat(proxystatus): dynamically generate XDS sync columns from discovered typeURLs for adaptive proxy-status output

* fix: unit tests

* fix lint target

* fix: release notes

* add a verbosity flag

* fix: release note

* fix: failing tests

* fix: relese note tests

* replace joinWithTabs with strings.Join

Signed-off-by: Aritra Dey <adey01027@gmail.com>

* better istioctl ps output

Signed-off-by: Aritra Dey <adey01027@gmail.com>

---------

Signed-off-by: Aritra Dey <adey01027@gmail.com>

howardjohn committed:
Initial support for ListenerSet (#55595)

* Initial ListenerSet support

* lint

* gen

* rebase

* Rebase

* rebase

* misc fixes

rudrakhp committed:
ensure consistency in inject label validation (#56737)

Signed-off-by: Rudrakh Panigrahi <rudrakh97@gmail.com>

ilrudie committed:
CNI Pod delete edge case (#56738)

* don't leave a dangling UID in the snapshot if we fail with a non-retryable

Signed-off-by: Ian Rudie <ian.rudie@solo.io>

* handle delete edge case

Signed-off-by: Ian Rudie <ian.rudie@solo.io>

* release note

Signed-off-by: Ian Rudie <ian.rudie@solo.io>

---------

Signed-off-by: Ian Rudie <ian.rudie@solo.io>

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56732)

DerekTBrown committed:
Add support for OPTIONAL_MUTUAL to Gateway API (#56742)

* implement OPTIONAL_MUTUAL conversion

* fmt

jaellio committed:
Remove ScopeByService on workload

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

keithmattix committed:
Multicluster AmbientIndex (#56395)

* Build global ambientindex implementation

Use nested join collections to store global versions of each
data structure that underpins the ambient index. We also create indexes
of these collections of collections based on the collection metadata so
that when can query, e.g. the Services from a particular cluster. This
commit also contains the first pass of the merging logic.

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Refactor multicluster collections

krt transformations can be called multiple times for the same inputm and
we were previously initializing new collections in every invocation.
Move to a common implementation of multicluster collections (based on
local and remote sub-collections) and a collection cache that ensures
only 1 collection per remote cluster exists in the global collection.

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Fix nestedjoin collection's seenFirstAddForKey

Previously, we used the collection as the synchronization
mechanism ensuring that we only get 1 add event if a key
gets added to multiple inner nested join collections at
the same time. This doesn't work because the krt event
handling framework has their own handlers and each handler
has its own dedicated goroutine, meaning that using the
collection to synchronize is too broad. Instead, in this
commit, we create a map + mutex data structure per handler,
preventing duplicate adds while ensuring that each handler's
view of the world is correct

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Add multicluster test cases to all existing ambient tests

Confirmed that these tests are all passing as of this commit. Had to
modify the ServiceEntry test slightly to deal with some ordering issues
where the workload entry would return first in the lookup

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Augment merge implementation with real olds

Update the nestedjoin merge implementation so that we can get real old
values. This prevents unnecessary xDS pushes up the stack

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Switch to mergedCache implementation

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Go back to setting collection metadata explicitly

Doing it in the krt options builder looks prettier, but it can lead to
subtle and widely impactful bugs where the builder's metadata is
unintentionally passed through to other clusters' collections.

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Always panic if cluster ID is not set in collection metadata

This will ensure we catch structural issues at compile/test time and we
don't run into tricky, silent bugs

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Check for config clustetr

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Become more robust against crashes

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Fixup test templates

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Use gw.Status for addresses

* Tests use gw.Status instead of Spec

* Add gatewayclass suffixe to default EW gateway SA

* Use status in AmbientSync test

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Fix lint

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* No, actually: become more resilient to crashes

Fix a rare data race that occurs when a cluster gets added to the
clusters collection before it's initialized. Also address several PR
comments.

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Regen golden files

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Fix lint

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

---------

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>
Co-authored-by: Steven Jin Xuan <sjinxuan@microsoft.com>

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56771)

jaellio committed:
Fix fmt

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

jaellio committed:
Add scope to serviceInfo equality check

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

german1608 committed:
Fix Gateway API tests to consider istio test cli flags (#56704)

* set revision when running istioctl

* Add revision to gateway in ManagedGatewayTest

* Do not hardcode istio-system in UnamagedGatewayTest

* include istioNamespace in istioctl commands for TaggedGatewayTest

* Use Revisions instead of Revision (later is deprecated)

* only set --istioNamespace is c.istioNamespace

* If ingress gateway ns not set, use istio-system

* set also ingress gateway svc name

* remove revision from ManagedGatewayTest (not needed), add TODO comment for remaining test

german1608 committed:
feat: represent revision tags using services (#56141)

* first iteration of using services and mutatingwebhooks as tags

* include namespaces

* log warnings on conflict

* rename revision-tags to revision-tags-mwc.yaml

* Create service tags on helm istiod chart

* initial iteration for istioctl tag set

* include IstioNamespace in GenerateOptions

* Update signature of Generate to return a list of objects to generate

* consider service if running ambient

* go back to strings when returning

* go back to string completely, delete TagResources

* rename service object

* move break to correct line

* implement istioctl tag remove

* add uniqTagsFromServices and uniqTagsFromWebhooks

* istioctl tag list consider services

* add operatorManageWebhooks guard to revision-tags-mwc.yaml

* rename revision-tags.yaml to revision-tags-svc.yaml

* refactor delete test case

* test istioctl tag list

* remove ambient check, create and manage both resources at the same time

* wip test cases

* remove controlplane mode

* fix test

* my own code review

* fix compilation error

* fix lint

* wip integ-helm_istio

* add app=istiod label filter

* merge override labels in generate.go

* add release note

* set owner labels to revision service tag

* rollback changes on e2e for helm

* fix helm e2e

* use servicesAfters instead of webhooksAfter in tag_test.go assertion log

* Remove unnecessary deprecation message

* refactor tag_test.go to reduce diff size

zirain committed:
Update client go (#56780)

* Automator: update istio/client-go@master dependency in istio/istio@master

* fix test

* fix more

* fix TestTelemetryMetricsExhaustiveness

---------

Co-authored-by: istio-testing <istio-testing-bot@google.com>

Stevenjin8 committed:
Messy checkpoint

Stevenjin8 committed:
Random

Stevenjin8 committed:
Jackie rebase

Stevenjin8 committed:
Jackie rebase

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56781)

zirain committed:
Impl RetryBudget in DestinationRule (#56782)

* impl RetryBudget

* release notes

* add test case

zirain committed:
add release notes for ClusterTrustBundle (#56748)

MaxBab committed:
Add domain resolve to Ambient TestServiceEntryInlinedWorkloadEntry test (#56784)

When executing TestServiceEntryInlinedWorkloadEntry Ambient test in a
public cloud based cluster, the created ingress is a domain based name
and the test fails with the following error:

configuration is invalid: endpoint address
"a6eec2a91721a42b0aa8e140fc6611fc-1873540778.us-east-1.elb.amazonaws.com"
is not a valid IP address

Add a domain name resolution to ip based address.

Signed-off-by: Maxim Babushkin <mbabushk@redhat.com>

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#56788)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56789)

howardjohn committed:
Do not transitively import net/trace (#56637)

* Do not transitively import net/trace

This drops 300kb off the binary. Note a big deal but
https://github.com/istio/istio/issues/56636 is a 5mb potentional and
this is a step to it

* drop license

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56790)

stevenctl committed:
do not use discovery filter for system namespace network (#56687)

* do not use discovery filter for system namespace network

* relnote

* revert label selector

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56794)

howardjohn committed:
status: fix busyloop (#56798)

This has been observed to cause extremely high CPU usage as we busyloop.

Basically if we have at least 1 pending item, and we re-enqueue and
actively worked on item, it will busy loop until the actively worked on
item is handled.

We should instead kill this worker. The other worker will handle the
task when it completes

howardjohn committed:
Gateway: provide stable HTTPRoute status ordering (#56504)

* Add test cases

* gateway: emit HTTPRoute status in a stable ordering

* Update tests

* fix test

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#56803)

jewertow committed:
manifests: refactor conditions for tproxy and capNetBindService (#56809)

* manifests: refactor conditions for TPROXY

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Encapsulate checking capNetBindService

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

---------

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

jewertow committed:
tests: assign privileged SCC to tproxy SA on OpenShift (#56799)

* tests: assign privileged SCC to tproxy SA on OpenShift

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Rename bool variable `openshift` to `isOpenShift`

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

* Fix a typo in CRB name

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

---------

Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

ramaraochavali committed:
fix dns althost name generation (#56806)

* fix dns althost name generation

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* remove unnecessary test change

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

---------

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali committed:
add gateway implementation for retry budget (#56808)

* add gateway implementation for retry budget

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* comments

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

---------

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

howardjohn committed:
tcp/tls route: fix multiple parent binds (#56818)

Fixes https://github.com/istio/istio/issues/56769

The problem was we had non-unique names for each generated object. Add
unit tests and fix to use the same counting appraoch HTTP/GRPC uses

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56813)

ramalabeysekera committed:
Add annotation support to deployment object in istio discovery chart (#56787)

* Add deployment object annotation support to istio discovery chart

* Add release note for deployment object annotation support to istiod chart

istio-testing committed:
Automator: update go-control-plane in istio/istio@master (#56821)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56822)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56823)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56829)

jaellio committed:
Create Istio owned CNI config (#56156)

* Create Istio owned CNI config

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Set CNIConfName to primary CNI config file name

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Make Istio owned cni config opt in and the config name
configurable

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Add warning for invalid istio own cni config filename priority

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

* Add release note

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

---------

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Stevenjin8 committed:
Merge branch 'steven/multicluster-ambient-index' of github.com:stevenjin8/istio into steven/multicluster-ambient-index

keithmattix committed:
Revert "feat: represent revision tags using services (#56141)" (#56833)

This reverts commit 0550a3176af1684bb7ea3b8fb133ec6957549071.

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#56835)

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#56840)

my-git9 committed:
improve istioctl dashboard example (#56842)

Signed-off-by: xin.li <xin.li@daocloud.io>

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#56834)

Stevenjin8 committed:
Tests passing (?)

Stevenjin8 committed:
Lint

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#56846)

Stevenjin8 committed:
Merge branch 'master' into steven/multicluster-ambient-index

Stevenjin8 committed:
Enable domain suffixes

Stevenjin8 committed:
fix tests again

Stevenjin8 committed:
Make gen

jaellio committed:
Add e/w gateway support for ambient multi network tests

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

jaellio committed:
Fix integration tests

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

jaellio committed:
Fix lint/gen and test failures

Signed-off-by: Jackie Elliott <jaellio@microsoft.com>

Stevenjin8 committed:
Merge remote-tracking branch 'jackie/jaellio/multicluster-ambient-index' into HEAD
